Add integration test for staging env w/ Puppeteer

LeoSL opened the pull request:
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/webhint.io)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

## Short description of the change(s)

### Fix #810

- Installs Puppeteer as a dev dependency in package.json.
- Requests a collection of webhint.io staging URLs and saves the response in a
Map() object.
- Iterates over the Map and checks if is there any status different from 200 or
another unknown error.
- If there is an error, throws it to interrupt the test execution

### 🕹 How to test

- [ ] Run the script `node helpers/integration-tests.js` ✅

- [ ] If you want to see the script finding an error you can add a bad URL to the URL collection.
Example: `http://pudim.com.br/xunda.jpg`

molant commented on helpers/integration-tests.js:
```diff
@@ -0,0 +1,53 @@
+                            .finally(() => {
+                                if (index === (URLsToVerify.length - 1)) {
+                                    resolve(responses);
+                                    browser.close();
```
[`browser.close()` returns a Promise](https://pptr.dev/#?product=Puppeteer&version=v1.20.0&show=api-pagecloseoptions).
I'm thinking of 2 possibilities, if you have a better one please go for it:

1. Add the condition after `responses.set()`. Although we might end up with a bit more nesting 🤔
1. Use `async/await` inside a `for..of` and test the pages one by one. This might make the code a bit easier even if it means the test will take a few more seconds (which is fine TBH). 

What do you think?

sarvaje replied:
I don't think we need to keep the responses in a `Map`.
As @molant says, you should use `async/await` inside a `for..of` and after each scan, print the status (that way we can see if it is failing more than one url) and after all the urls are scanned, throw and error if there was any exception (or maybe it is enough if we return an exitCode !== 0)

LeoSL commented:
> Also, IMO, we should check that running a scan is working too.

Can you please clarify what you mean by "check that running a scan is working"?

sarvaje commented:
>@sarvaje Can you please clarify what you mean by "check that running a scan is working"?

Sorry hehehe. I mean going to the `/scanner` and some url `e.g. https://example.com` click on `run scan` and see if you get a status code 200.

Some times, everything looks good, but then when you scan an url, something is broken.

molant commented:
There are a couple methods in puppeteer that make this relatively easy:

* [page.type](https://pptr.dev/#?product=Puppeteer&version=v1.20.0&show=api-pagetypeselector-text-options): it accepts a css selector and the text to type on the input
* [page.click](https://pptr.dev/#?product=Puppeteer&version=v1.20.0&show=api-pageclickselector-options) which accepts a css selector and will "click" on it

Because we are going to navigate to a page we will probably have to use [`page.waitForNavigation`](https://pptr.dev/#?product=Puppeteer&version=v1.20.0&show=api-pagewaitfornavigationoptions) to make sure that we navigate to a results page and get a 200 status code. The docs page has a good example on how to use `waitForNavigation` in combination with `click` and such but I think it will be something similar to:

```js
await page.goto('/scanner/');
await page.type('selector for input', 'url to analyze');
const [result] = await Promise.all([
  page.waitForNavigation(),
  page.click('selector for button')
]);

// check result here
```

LeoSL commented:
@molant I've rebased from upstream master and did the following:

- Refactored the code to use async/await syntax sugar instead of Promise chains
- Isolated the scanner test

Let me know if the changes sound good to you.
Thanks!

molant commented on helpers/integration-tests_deprecated.js:
```diff
@@ -0,0 +1,53 @@
+const puppeteer = require('puppeteer');
```
We no longer need this file 🙂

LeoSL replied:
Oh, I forgot that one! Thanks!

molant replied:
It looks like this file is still in the PR

molant commented on helpers/integration-tests/staticUrls.js:
```diff
@@ -0,0 +1,45 @@
+    return new Promise((resolve) => {
+        let errorFound = false;
+
+        staticURLsToVerify.forEach(async (url, index) => {
```
I believe `.forEach` doesn't `await` so all the items in the array are called at the same time and thus starting multiple different instances of the browser. When testing this how many separate chrome windows did you see? Was it just one being opened and closed or several?

If you saw multiple instances then we should probably change this to something like this:

```js
const runStaticTests = async () => {
    let errorFound = false;

    for (const url of staticURLsToVerify) {
         const resultStatus = await runPuppeteer(url);
        
        if (resultStatus !== 200) {
            errorFound = true;
        }
    }

    return errorFound;
}
```

LeoSL replied:
@molant I'm running the Chrome headless version so I don't observe the windows popping up - thus I can't answer your question about the instances.

When running on the CLI they appear pretty much sequential on all my tests. I'm assuming that if they're being fired sequentially the printing (console.log) order would change on ad hoc basis.

molant replied:
As I suspected [`forEach` doesn't `await`](https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop) , I've been bitten by this in the past 😅

With the code I posted earlier you don't need access to the index. Because the urls are being analyzed sequentially, as soon as the `for..of` ends you can return `errorFound`. We also add the keyword `async` so no need to create a `Promise` object.

LeoSL replied:
That's great, thanks @molant 
I will do the fixes later today! 👍

LeoSL replied:
@molant Done!

LeoSL commented:
Sorry, @molant forgot again to remove. Now I believe everything looks fine!

sarvaje commented on helpers/integration-tests/scanner.js:
```diff
@@ -0,0 +1,26 @@
+        }! => Status: ${response.status()}`
+    );
+
+    return response.status() !== 200;
```
NIT: In `staticUrls.js` you have this code wrapped in a `try/catch` maybe it would be nice to do the same here.

antross commented:
@molant Looks like we two approvals now. Is there anything else you wanted or is this good to merge?

molant commented:
Good to merge. Haven't had the time to merge today. Can you please take care of it?


________________________________
From: Tony Ross <notifications@github.com>
Sent: Monday, October 14, 2019 13:20
To: webhintio/webhint.io
Cc: Antón Molleda; Mention
Subject: Re: [webhintio/webhint.io] Add integration test for staging env w/ Puppeteer (#843)


@molant<https://github.com/molant> Looks like we two approvals now. Is there anything else you wanted or is this good to merge?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/webhintio/webhint.io/pull/843?email_source=notifications&email_token=AAEUDATVVJGTGFPJRANNVBTQOTIBFA5CNFSM4I5YVAX2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBGLQWA#issuecomment-541898840>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAEUDAVEGYOAST5C2AG376DQOTIBFANCNFSM4I5YVAXQ>.
